zio-ftp/src/main/scala/zio/ftp/UnsecureFtp.scala

```diff
@@ -30,24 +32,58 @@ import zio.ZIO.{ acquireRelease, attemptBlockingIO }
  */
 final private class UnsecureFtp(unsafeClient: Client) extends FtpAccessors[Client] {
 
+  private var pendingExit: Option[Exit[IOException, Unit]] = None
```

Comment:
i dont think it is safe to have a mutable share state

Reply:
You can put it into an atomic reference--just to make it thread safe, mind you.

As long as the only access to this is through methods returning `ZIO`, it's still as 'purely functional' as anything else and won't inhibit retries, etc.

Reply:
Hey @jdegoes,

thanks for commenting. Maybe I'm missing something here, but this thing being essentially just a wrapper around a socket connection, isn't it completely thread unsafe anyway?

Reply:
Thanks again i was able to take some time and to read your changes, 

```
But it leads to another problem: finalizers aren't allowed to throw errors, but completePendingCommand can fail. I could die in that situation, but that doesn't seem right, because die indicates a defect in the program, and when the FTP server decides to be weird, that's not the program's fault. Therefore, I decided to not throw the error from the finalizer but instead delay the error until the next execute call.
```

I would not delay the error, since we are implementing a default finalizer to close the stream, i would only print the error 
```
ZIO.logError()
```

There is nothing we can do about it. 

Also the delay create more confusion, since you will not know that a previous error failed.
But doing this you can simplify  remove 
```
 private var pendingExit: Option[Exit[IOException, Unit]] = None
```

Reply:
this version could fix your issue

```scala
def readFile(path: String, chunkSize: Int = 2048, fileOffset: Long): ZStream[Any, IOException, Byte] = {
    val initialize = execute(_.setRestartOffset(fileOffset))

    val terminate: UIO[Unit] = ZIO
      .fail(
        FileTransferIncompleteError(s"Cannot finalize the file transfer and completely read the entire file $path.")
      )
      .unlessZIO(execute(_.completePendingCommand()))
      .option
      .unit

    val inputStream =
      execute(c => Option(c.retrieveFileStream(path))).someOrFail(InvalidPathError(s"File does not exist $path"))

    ZStream.unwrap(
      for {
        _ <- initialize
        is <- inputStream
      } yield ZStream.fromInputStream(is, chunkSize).ensuring(terminate)
    )
  }
```

Reply:
Hey @regis-leray ,

I understand your concerns about delaying error propagation, I agree it's not ideal. I did it this way because I wanted to keep the API kind-of-compatible while still exposing the error if it occurs.

But I feel uneasy about the `.option.unit` thing too because then there is no possibility at all to programmatically detect that an error has happened.

Maybe we could try a loan-pattern style API here?
```
def readFile[R, E >: IOException, A](
  f: ZStream[Any, IOException, Byte] => ZIO[R, E, A]
): ZIO[R, E, A]
```
The implementation would work like this:
 - create the stream
 - process it with the `f` function
 - when done, call `completePendingCommand`
 - if either `f` or `completePendingCommand` fails, the error is raised to the caller.
 - if they both fail, the error raised by `f` is returned to the caller and the `completePendingCommand` error is logged and ignored
What do you think?

Also, I'd like to retain at least some amount of mutable state in order to detect the situation where the user is trying to stuff with the FTP client before the stream is finalized, because that is never going to work and it  currently leads to errors that are just wrong, like the "file not found" error I've had.